deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl

```diff
@@ -2266,7 +2279,41 @@ publish_to_queues_with_checks(
                     Error
             end;
         {error, access_refused} ->
-            {error, access_refused, State}
+            %%  If disconnect_on_unauthorized is false,
+            %%  MQTT v5 and QoS1 reply with PUBACK including an error reason code and keep connection,
+            %%  MQTT v3 and QoS1 reply with PUBACK no error reason code and keep connection,
+            %%  QoS0 drop silently and keep connection.
+            %%  Otherwise, disconnect.
+            case application:get_env(rabbitmq_mqtt, disconnect_on_unauthorized, true) of
+                false ->
+                    case Msg#mqtt_msg.qos of
+                        ?QOS_1 ->
+                            case State#state.cfg#cfg.proto_ver of
+                                ?MQTT_PROTO_V5 ->
+                                    Reply = #mqtt_packet{
+                                        fixed = #mqtt_packet_fixed{type = ?PUBACK},
+                                        variable = #mqtt_packet_puback{
+                                            packet_id = Msg#mqtt_msg.packet_id,
+                                            reason_code = ?RC_NOT_AUTHORIZED
+                                        }
+                                    },
+                                    _ = send(Reply, State);
+                                _ ->
+                                    Reply = #mqtt_packet{
+                                        fixed = #mqtt_packet_fixed{type = ?PUBACK},
+                                        variable = #mqtt_packet_puback{
+                                            packet_id = Msg#mqtt_msg.packet_id
+                                        }
+                                    },
+                                    _ = send(Reply, State)
+                            end;
```

Comment:
These two separate branches are not necessary. Module `rabbit_mqtt_packet` encodes the packet correctly without the reason code for MQTT v4 and v3.